lib_eio/semaphore.ml

```diff
@@ -1,55 +1,42 @@
-type state =
-  | Free of int
-  | Waiting of unit Waiters.t
-
 type t = {
   id : Ctf.id;
-  mutex : Mutex.t;
-  mutable state : state;
+  state : Sem_state.t;
 }
 
 let make n =
-  if n < 0 then raise (Invalid_argument "n < 0");
   let id = Ctf.mint_id () in
   Ctf.note_created id Ctf.Semaphore;
   {
     id;
-    mutex = Mutex.create ();
-    state = Free n;
+    state = Sem_state.create n;
   }
 
 let release t =
-  Mutex.lock t.mutex;
   Ctf.note_signal t.id;
-  match t.state with
-  | Free x when x = max_int -> Mutex.unlock t.mutex; raise (Sys_error "semaphore would overflow max_int!")
-  | Free x -> t.state <- Free (succ x); Mutex.unlock t.mutex
-  | Waiting q ->
-    begin match Waiters.wake_one q () with
-      | `Ok -> ()
-      | `Queue_empty -> t.state <- Free 1
-    end;
-    Mutex.unlock t.mutex
+  Sem_state.release t.state
 
-let rec acquire t =
-  Mutex.lock t.mutex;
-  match t.state with
-  | Waiting q ->
-    Ctf.note_try_read t.id;
-    Waiters.await ~mutex:(Some t.mutex) q t.id
-  | Free 0 ->
-    t.state <- Waiting (Waiters.create ());
-    Mutex.unlock t.mutex;
-    acquire t
-  | Free n ->
-    Ctf.note_read t.id;
-    t.state <- Free (pred n);
-    Mutex.unlock t.mutex
+let acquire t =
+  if not (Sem_state.acquire t.state) then (
+    (* No free resources.
+       We must wait until one of the existing users increments the counter and resumes us.
+       It's OK if they resume before we suspend; we'll just pick up the token they left. *)
+    Suspend.enter_unchecked (fun ctx enqueue ->
+        match Sem_state.suspend t.state (fun () -> enqueue (Ok ())) with
```

Comment:
Is there a potential here to avoid the `Suspend.enter_unchecked` call (which, I assume, captures the continuation) by exposing the `Sem_state` protocol a bit more.  Basically, require the acquirer to separately create the cell.  After creating the cell, the acquirer would then read the cell state (as is done in `Sem_state.suspend` at the moment) before capturing the continuation.  If the state has already been set to `Finished`, then there is no need to capture a continuation.  Otherwise capture continuation and proceed as before.  Basically, this results in a kind of double-checked pattern.  This could improve performance in highly contested cases where it is possible that a resumer actually manages to see the `InTransition` (or `Empty`) state.

Reply:
I don't think so. This case wouldn't work:

1. The suspender decrements the count to say they're planning to suspend.
2. The resumer resumes the cell.
3. The suspender creates the cell.

We have to be able to initialise cells before the suspender does anything beyond modifying the counter. We don't want it creating cells before changing the counter because that includes the fast path (where we don't need a cell at all).

Reply:
Just to be clear, this is what I had in mind:

```diff
diff --git a/lib_eio/sem_state.ml b/lib_eio/sem_state.ml
index 24a301c..7ffebf6 100644
--- a/lib_eio/sem_state.ml
+++ b/lib_eio/sem_state.ml
@@ -109,6 +109,26 @@ let acquire t =
      which happens if we decremented *from* a positive one. *)
   s > 0
 
+let prepare_suspend t =
+  Cells.next_suspend t.cells
+
+let is_in_transition (_, (cell: cell Atomic.t)) =
+  match Atomic.get cell with
+  | In_transition -> true
+  | Finished | Request _ -> false
+
+let perform_suspend t (segment, (cell: cell Atomic.t)) k : request option =
+  if Atomic.compare_and_set cell In_transition (Request k) then Some (t, segment, cell)
+  else match Atomic.get cell with
+    | Finished ->
+      (* We got resumed before we could add the waiter. *)
+      k ();
+      None
+    | Request _ | In_transition ->
+      (* These are unreachable from the previously-observed non-In_transition state
+         without us taking some action first *)
+      assert false
+
 let suspend t k : request option =
   let (segment, cell) = Cells.next_suspend t.cells in
   if Atomic.compare_and_set cell In_transition (Request k) then Some (t, segment, cell)
diff --git a/lib_eio/semaphore.ml b/lib_eio/semaphore.ml
index 2733be5..a90fa64 100644
--- a/lib_eio/semaphore.ml
+++ b/lib_eio/semaphore.ml
@@ -20,21 +20,24 @@ let acquire t =
     (* No free resources.
        We must wait until one of the existing users increments the counter and resumes us.
        It's OK if they resume before we suspend; we'll just pick up the token they left. *)
-    Suspend.enter_unchecked (fun ctx enqueue ->
-        match Sem_state.suspend t.state (fun () -> enqueue (Ok ())) with
-        | None -> ()   (* Already resumed *)
-        | Some request ->
-          Ctf.note_try_read t.id;
-          match Fiber_context.get_error ctx with
-          | Some ex ->
-            if Sem_state.cancel request then enqueue (Error ex);
-            (* else already resumed *)
-          | None ->
-            Fiber_context.set_cancel_fn ctx (fun ex ->
-                if Sem_state.cancel request then enqueue (Error ex)
-                (* else already resumed *)
-              )
-      )
+    let segment_cell = Sem_state.prepare_suspend t.state in
+    if Sem_state.is_in_transition segment_cell then (
+      Suspend.enter_unchecked (fun ctx enqueue ->
+          match Sem_state.perform_suspend t.state segment_cell (fun () -> enqueue (Ok ())) with
+          | None -> ()   (* Already resumed *)
+          | Some request ->
+            Ctf.note_try_read t.id;
+            match Fiber_context.get_error ctx with
+            | Some ex ->
+              if Sem_state.cancel request then enqueue (Error ex);
+              (* else already resumed *)
+            | None ->
+              Fiber_context.set_cancel_fn ctx (fun ex ->
+                  if Sem_state.cancel request then enqueue (Error ex)
+                  (* else already resumed *)
+                )
+        )
+    )
   );
   Ctf.note_read t.id
```

Reply:
Ah, I see. Pushing some of the work to outside of the suspend in case we can skip the suspend by the time it's done, at the cost of an extra `Atomic.get`.

I tried measuring it by incrementing `slow` and `fast` counters in `is_in_transition` and running the benchmark gives:
```
slow=850475, fast=75, frac=0.01%
```
So probably not worth it, I think.

Reply:
OK, I guess I should have disabled the single-process runs before measuring that... it's `slow=51221, fast=98, frac=0.19%` now (0.19% of the slow path cases). But I don't think that changes anything.